aten/src/ATen/native/TensorIterator.cpp

```diff
@@ -402,9 +402,11 @@ void TensorIterator::compute_types(const TensorIteratorConfig& config) {
       // Casts to outputs by creating temporaries of the correct dtype (if needed)
       if (config.cast_common_dtype_to_outputs_ && op.is_output && op.current_dtype != common_dtype_) {
         op.original_tensor = op.tensor;
-        op.tensor = at::empty_like(op.tensor,
-                                   op.tensor.options().dtype(common_dtype_),
-                                   LEGACY_CONTIGUOUS_MEMORY_FORMAT);
+        AutoNonVariableTypeMode non_variable_type_mode;
```

Comment:
The AutoNonVariableTypeMode additions here are super sketchy. This is related to the Q you asked earlier in composability chat:

> I've submitted https://github.com/pytorch/pytorch/pull/43773 where I make TensorIterator use at::native::empty instead of at::empty, to avoid dispatch costs. I had test failures related to autograd tests for logit function. logit derivative is defined as 
> - name: logit(Tensor self, float? eps=None) -> Tensor
  self: "GradMode::is_enabled() ? infinitely_differentiable_logit_backward(grad, self, eps) : logit_backward(grad, self, eps)"
> 
> and in logit_backward I was getting 
> RuntimeError: impl::variable_excluded_from_dispatch() INTERNAL ASSERT FAILED at "../aten/src/ATen/native/TensorFactories.cpp":182, please report a bug to PyTorch. 
Exception raised from empty at ../aten/src/ATen/native/TensorFactories.cpp:182 (most recent call first):
>
> I worked around it by adding not_implemented gradients for logit_backward function. It's still weird that I had to do that, because logit_backward is called only when GradMode is not enabled. Interestingly, silu and prelu were not affected and did not require this workaround, even though their derivatives are similar, and gradients for silu_backward are not defined:
> - name: silu(Tensor self) -> Tensor
>   self: "GradMode::is_enabled() ? infinitely_differentiable_silu_backward(grad, self) : silu_backward(grad, self)"
>
> Any idea why that would be? One thing that's different between them is logit is exposed as torch function and method, silu is accessible only via torch.nn.functional. 
> My PR brings ~10% speed-up for out-of-place operations overheads 

I'd definitely like to get a good answer about why we are calling TensorIterator without autograd mode having already been disabled; my preference would have been for us to fix those cases instead of directly bashing them in TensorIterator itself.

Reply:
This is happening in the tests https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/test/tensor_iterator_test.cpp#L54-L56. I'm guessing it also could happen if someone were to use TensorIterator in custom op. 
To add to this, DeviceGuard business is also sketchy, we should be calling TensorIterator with proper deviceGuard already, and yet in scalar-tensor case we don't (long story). And again, same thing can happen with people writing custom ops, they won't have properly set device guards.

Reply:
Oh ok, this makes a good deal of sense.

For the custom op situation, @bhosmer and @ailzhang recently agreed with me that in the long term we should have a fallback autograd kernel which excludes all dispatch keys, which would make custom ops work correctly. I suppose you are right that without these calls, in the short term you would break all custom ops using tensor iterator without these guards.

So OK, I'm fine with these guards, but I would appreciate it if we had a comment saying why the guards are needed and under what circumstances they could be removed.